include/aspect/freesurface.h

```diff
@@ -0,0 +1,246 @@
+       * of the domain.
+       */
+      void detach_manifolds();
+
+      /**
+       * Declare parameters for the free surface handling.
+       */
+      static
+      void declare_parameters (ParameterHandler &prm);
+
+      /**
+       * Parse parameters for the free surface handling.
+       */
+      void parse_parameters (ParameterHandler &prm);
+
+    private:
+      /**
+       * Set the boundary conditions for the solution of the elliptic
+       * problem, which computes the displacements of the internal
+       * vertices so that the mesh does not become too distored due to
+       * motion of the free surface.  Velocities of vertices on the free
+       * surface are set to be the normal of the Stokes velocity solution
+       * projected onto that surface.  Velocities of vertices on free-slip
+       * boundaries are constrained to be tangential to those boundaries.
+       * Velocities of vertices on no-slip boundaries are set to be zero.
+       */
+      void make_constraints ();
+
+      /**
+       * Project the the Stokes velocity solution onto the
+       * free surface. Called by make_constraints()
+       */
+      void project_velocity_onto_boundary (LinearAlgebra::Vector &output);
+
+      /**
+       * Solve vector Laplacian equation for internal mesh displacements.
+       */
+      void compute_mesh_displacements ();
+
+      /**
+       * Calculate the velocity of the mesh for ALE corrections.
+       */
+      void interpolate_mesh_velocity ();
+
+      /**
+       * Reference to the Simulator object to which a FreeSurfaceHandler
+       * instance belongs.
+       */
+      Simulator<dim> &sim;
+
+      /**
+      * Finite element for the free surface implementation, which is
+      used for tracking mesh deformation.
+      */
+      const FESystem<dim> free_surface_fe;
+
+      /**
+       * DoFHanlder for the free surface implementation.
+       */
+      DoFHandler<dim> free_surface_dof_handler;
+
+
+      /**
+       * Stabilization parameter for the free surface.  Should be between
+       * zero and one. A value of zero means no stabilization.  See Kaus
+       * et. al. 2010 for more details.
+       */
+      double free_surface_theta;
+
+      /**
+       * BlockVector which stores the mesh velocity.
+       * This is used for ALE corrections.
+       */
+      LinearAlgebra::BlockVector mesh_velocity;
+
+      /**
+       * Vector for storing the positions of the mesh vertices. This
+       * is used for calculating the mapping from the reference cell to
+       * the position of the cell in the deformed mesh. This must be
+       * redistributed upon mesh refinement.
+       */
+      LinearAlgebra::Vector mesh_displacements;
+
+      /**
+       * Vector for storing the mesh velocity in the free surface finite
+       * element space, which is, in general, not the same finite element
+       * space as the Stokes system.
+       */
+      LinearAlgebra::Vector fs_mesh_velocity;
```

Comment:
Do you actually still need this as a member variable? It would seem to me that this is now just a temporary variable in the function that computes the mesh velocity, but that a multiple of it is then added to `mesh_displacement`. If so, I'd prefer if it were made a variable in that function, so that one is not tempted to use it elsewhere (which at this point would almost certainly be a mistake, since we don't want to use it any more.)